File: Stacks and Queues/PriorityQueue.c
#include <stdio.h>
#include <stdlib.h>

#define SIZE 5  // Define the maximum size of the priority queue

// Structure to store elements and their priorities
struct PriorityQueue {
    int items[SIZE];
    int priorities[SIZE];
    int count;
};

// Initialize the priority queue
void initializeQueue(struct PriorityQueue* pq) {
    pq->count = 0;  // Set initial count to 0
}

// Check if the priority queue is full
int isFull(struct PriorityQueue* pq) {
    return pq->count == SIZE;
}

// Check if the priority queue is empty
int isEmpty(struct PriorityQueue* pq) {
    return pq->count == 0;
}

// Enqueue operation with priority
void enqueue(struct PriorityQueue* pq, int value, int priority) {
    if (isFull(pq)) {
        printf("Queue is full!\n");
        return;
    }

    int i;
    // Find the correct position based on priority
    for (i = pq->count - 1; i >= 0 && pq->priorities[i] < priority; i--) {
        pq->items[i + 1] = pq->items[i];
        pq->priorities[i + 1] = pq->priorities[i];
    }
    // Insert the element in its position
    pq->items[i + 1] = value;
    pq->priorities[i + 1] = priority;
    pq->count++;
    printf("Inserted %d with priority %d\n", value, priority);
}

// Dequeue operation (removes the highest priority element)
int dequeue(struct PriorityQueue* pq) {
    if (isEmpty(pq)) {
        printf("Queue is empty!\n");
        return -1;
    }
    // Remove the element with the highest priority (first element)
    int value = pq->items[0];
    // Shift all remaining elements
    for (int i = 0; i < pq->count - 1; i++) {
        pq->items[i] = pq->items[i + 1];
        pq->priorities[i] = pq->priorities[i + 1];
    }
    pq->count--;
    printf("Deleted %d\n", value);
    return value;
}

// Function to display the queue
void display(struct PriorityQueue* pq) {
    if (isEmpty(pq)) {
        printf("Queue is empty!\n");
    } else {
        printf("Queue elements with priorities: ");
        for (int i = 0; i < pq->count; i++) {
            printf("%d(p%d) ", pq->items[i], pq->priorities[i]);
        }
        printf("\n");
    }
}

int main() {
    struct PriorityQueue pq;
    initializeQueue(&pq);

    enqueue(&pq, 10, 2);
    enqueue(&pq, 20, 1);
    enqueue(&pq, 30, 3);
    enqueue(&pq, 40, 0);
    
    display(&pq);  // Display the queue
    
    dequeue(&pq);  // Remove the highest priority element
    dequeue(&pq);  // Remove the next highest priority element
    
    display(&pq);  // Display the queue after deletion
    
    return 0;
}

File: Stacks and Queues/PriorityQueue.py
class PriorityQueue:
    def __init__(self):
        """Initialize an empty priority queue"""
        self.queue = []

    def is_empty(self):
        """Check if the queue is empty"""
        return len(self.queue) == 0

    def enqueue(self, value, priority):
        """Insert an element into the priority queue with a given priority"""
        # Insert the element as a tuple (priority, value)
        self.queue.append((priority, value))
        # Sort the queue by priority (ascending order)
        self.queue.sort(key=lambda x: x[0])
        print(f"Inserted {value} with priority {priority}")

    def dequeue(self):
        """Remove and return the element with the highest priority"""
        if self.is_empty():
            print("Queue is empty!")
            return None
        # Remove the element with the highest priority (lowest priority number)
        element = self.queue.pop(0)
        print(f"Deleted {element[1]}")
        return element[1]

    def display(self):
        """Display the elements of the priority queue"""
        if self.is_empty():
            print("Queue is empty!")
        else:
            print("Queue elements with priorities:", end=" ")
            for priority, value in self.queue:
                print(f"{value}(p{priority})", end=" ")
            print()


# Test the PriorityQueue class
if __name__ == "__main__":
    pq = PriorityQueue()

    pq.enqueue(10, 2)
    pq.enqueue(20, 1)
    pq.enqueue(30, 3)
    pq.enqueue(40, 0)

    pq.display()  # Display the queue
    
    pq.dequeue()  # Remove the highest priority element
    pq.dequeue()  # Remove the next highest priority element
    
    pq.display()  # Display the queue after deletion
